flocker/node/agents/blockdevice.py

```diff
@@ -53,6 +59,125 @@ class UnattachedVolume(VolumeException):
     """
 
 
+DATASET = Field(
+    u"dataset",
+    lambda dataset: dataset.dataset_id,
+    u"The unique identifier of a dataset."
+)
+
+DATASET_ID = Field(
+    u"dataset_id",
+    lambda dataset_id: unicode(dataset_id),
+    u"The unique identifier of a dataset."
+)
+
+MOUNTPOINT = Field(
+    u"mountpoint",
+    lambda path: path.path,
+    u"The absolute path to the location on the node where the dataset will be "
+    u"mounted.",
+)
+
+DEVICE_PATH = Field(
+    u"block_device_path",
+    lambda path: path.path,
+    u"The absolute path to the block device file on the node where the "
+    u"dataset is attached.",
+)
+
+BLOCK_DEVICE_ID = Field(
+    u"block_device_id",
+    lambda id: unicode(id),
+    u"The unique identifier if the underlying block device."
+)
+
+BLOCK_DEVICE_SIZE = Field(
+    u"block_device_size",
+    lambda size: size,
+    u"The size of the underlying block device."
+)
+
+BLOCK_DEVICE_HOST = Field(
+    u"block_device_host",
+    lambda host: host,
+    u"The host to which the underlying block device is attached."
+)
+
+CREATE_BLOCK_DEVICE_DATASET = ActionType(
+    u"agent:blockdevice:create",
+    [DATASET, MOUNTPOINT],
+    [DEVICE_PATH, BLOCK_DEVICE_ID, DATASET_ID, BLOCK_DEVICE_SIZE,
+     BLOCK_DEVICE_HOST],
+    u"A block-device-backed dataset is being created.",
+)
+
+
+@implementer(IStateChange)
+class CreateBlockDeviceDataset(PRecord):
+    """
```

Comment:
When you break out the components into sub-actions, I recommend doing the thing we discussed where `CreateBlockDeviceDataset` continues to exist... but is a compound `IStateChange` that runs sub-`IStateChange` instances.

That means high-level tests can still just assert "if dataset isn't created then do this high-level action", and you can have separate test that is "this high-level action is composed of these sub-actions".